<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Introduction to Biomedical Data Science</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="This book introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biomedical data.">
  <meta name="generator" content="bookdown 0.3 and GitBook 2.6.7">

  <meta property="og:title" content="Introduction to Biomedical Data Science" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biomedical data." />
  <meta name="github-repo" content="bioconnector/workshops" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Introduction to Biomedical Data Science" />
  
  <meta name="twitter:description" content="This book introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biomedical data." />
  

<meta name="author" content="Stephen Turner &amp; VP Nagraj">


<meta name="date" content="2017-02-11">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="reproducible-research.html">
<link rel="next" href="surival-analysis.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to Biomedical Data Science</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="r-basics.html"><a href="r-basics.html"><i class="fa fa-check"></i><b>2</b> R Basics</a><ul>
<li class="chapter" data-level="2.1" data-path="r-basics.html"><a href="r-basics.html#rstudio"><i class="fa fa-check"></i><b>2.1</b> RStudio</a></li>
<li class="chapter" data-level="2.2" data-path="r-basics.html"><a href="r-basics.html#basic-operations"><i class="fa fa-check"></i><b>2.2</b> Basic operations</a></li>
<li class="chapter" data-level="2.3" data-path="r-basics.html"><a href="r-basics.html#functions"><i class="fa fa-check"></i><b>2.3</b> Functions</a></li>
<li class="chapter" data-level="2.4" data-path="r-basics.html"><a href="r-basics.html#data-frames"><i class="fa fa-check"></i><b>2.4</b> Data Frames</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="data-frames-1.html"><a href="data-frames-1.html"><i class="fa fa-check"></i><b>3</b> Data Frames</a><ul>
<li class="chapter" data-level="3.1" data-path="data-frames-1.html"><a href="data-frames-1.html#our-data"><i class="fa fa-check"></i><b>3.1</b> Our data</a></li>
<li class="chapter" data-level="3.2" data-path="data-frames-1.html"><a href="data-frames-1.html#reading-in-data"><i class="fa fa-check"></i><b>3.2</b> Reading in data</a><ul>
<li class="chapter" data-level="3.2.1" data-path="data-frames-1.html"><a href="data-frames-1.html#dplyr-and-readr"><i class="fa fa-check"></i><b>3.2.1</b> dplyr and readr</a></li>
<li class="chapter" data-level="3.2.2" data-path="data-frames-1.html"><a href="data-frames-1.html#read_csv"><i class="fa fa-check"></i><b>3.2.2</b> <code>read_csv()</code></a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="data-frames-1.html"><a href="data-frames-1.html#inspecting-data.frame-objects"><i class="fa fa-check"></i><b>3.3</b> Inspecting data.frame objects</a></li>
<li class="chapter" data-level="3.4" data-path="data-frames-1.html"><a href="data-frames-1.html#accessing-variables-subsetting-data-frames"><i class="fa fa-check"></i><b>3.4</b> Accessing variables &amp; subsetting data frames</a></li>
<li class="chapter" data-level="3.5" data-path="data-frames-1.html"><a href="data-frames-1.html#bonus-preview-to-advanced-manipulation"><i class="fa fa-check"></i><b>3.5</b> BONUS: Preview to advanced manipulation</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation with <code>dplyr</code></a><ul>
<li class="chapter" data-level="4.1" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#review"><i class="fa fa-check"></i><b>4.1</b> Review</a><ul>
<li class="chapter" data-level="4.1.1" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#our-data-1"><i class="fa fa-check"></i><b>4.1.1</b> Our data</a></li>
<li class="chapter" data-level="4.1.2" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#reading-in-data-1"><i class="fa fa-check"></i><b>4.1.2</b> Reading in data</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#the-dplyr-package"><i class="fa fa-check"></i><b>4.2</b> The dplyr package</a></li>
<li class="chapter" data-level="4.3" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#dplyr-verbs"><i class="fa fa-check"></i><b>4.3</b> dplyr verbs</a><ul>
<li class="chapter" data-level="4.3.1" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#filter"><i class="fa fa-check"></i><b>4.3.1</b> filter()</a></li>
<li class="chapter" data-level="4.3.2" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#select"><i class="fa fa-check"></i><b>4.3.2</b> select()</a></li>
<li class="chapter" data-level="4.3.3" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#mutate"><i class="fa fa-check"></i><b>4.3.3</b> mutate()</a></li>
<li class="chapter" data-level="4.3.4" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#arrange"><i class="fa fa-check"></i><b>4.3.4</b> arrange()</a></li>
<li class="chapter" data-level="4.3.5" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#summarize"><i class="fa fa-check"></i><b>4.3.5</b> summarize()</a></li>
<li class="chapter" data-level="4.3.6" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#group_by"><i class="fa fa-check"></i><b>4.3.6</b> group_by()</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#the-pipe"><i class="fa fa-check"></i><b>4.4</b> The pipe: <strong>%&gt;%</strong></a><ul>
<li class="chapter" data-level="4.4.1" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#how-works"><i class="fa fa-check"></i><b>4.4.1</b> How %&gt;% works</a></li>
<li class="chapter" data-level="4.4.2" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#nesting-versus"><i class="fa fa-check"></i><b>4.4.2</b> Nesting versus %&gt;%</a></li>
<li class="chapter" data-level="4.4.3" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#piping-exercises"><i class="fa fa-check"></i><b>4.4.3</b> Piping exercises</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="data-manipulation-with-dplyr.html"><a href="data-manipulation-with-dplyr.html#homework"><i class="fa fa-check"></i><b>4.5</b> Homework</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html"><i class="fa fa-check"></i><b>5</b> Tidy Data and <code>tidyr</code></a><ul>
<li class="chapter" data-level="5.1" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#review-1"><i class="fa fa-check"></i><b>5.1</b> Review</a><ul>
<li class="chapter" data-level="5.1.1" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#prior-classes"><i class="fa fa-check"></i><b>5.1.1</b> Prior classes</a></li>
<li class="chapter" data-level="5.1.2" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#data-needed"><i class="fa fa-check"></i><b>5.1.2</b> Data needed</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#tidy-data"><i class="fa fa-check"></i><b>5.2</b> Tidy data</a></li>
<li class="chapter" data-level="5.3" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#the-tidyr-package"><i class="fa fa-check"></i><b>5.3</b> The <strong>tidyr</strong> package</a><ul>
<li class="chapter" data-level="5.3.1" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#gather"><i class="fa fa-check"></i><b>5.3.1</b> <code>gather()</code></a></li>
<li class="chapter" data-level="5.3.2" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#separate"><i class="fa fa-check"></i><b>5.3.2</b> <code>separate()</code></a></li>
<li class="chapter" data-level="5.3.3" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#it-all-together"><i class="fa fa-check"></i><b>5.3.3</b> <code>%&gt;%</code> it all together</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#tidy-the-yeast-data"><i class="fa fa-check"></i><b>5.4</b> Tidy the yeast data</a><ul>
<li class="chapter" data-level="5.4.1" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#separate-the-name"><i class="fa fa-check"></i><b>5.4.1</b> <code>separate()</code> the <code>NAME</code></a></li>
<li class="chapter" data-level="5.4.2" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#gather-the-data"><i class="fa fa-check"></i><b>5.4.2</b> <code>gather()</code> the data</a></li>
<li class="chapter" data-level="5.4.3" data-path="tidy-data-and-tidyr.html"><a href="tidy-data-and-tidyr.html#inner_join-to-go"><i class="fa fa-check"></i><b>5.4.3</b> <code>inner_join()</code> to GO</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html"><i class="fa fa-check"></i><b>6</b> Data Visualization with <code>ggplot2</code></a><ul>
<li class="chapter" data-level="6.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#review-2"><i class="fa fa-check"></i><b>6.1</b> Review</a><ul>
<li class="chapter" data-level="6.1.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#gapminder-data"><i class="fa fa-check"></i><b>6.1.1</b> Gapminder data</a></li>
<li class="chapter" data-level="6.1.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#dplyr-review"><i class="fa fa-check"></i><b>6.1.2</b> dplyr review</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#about-ggplot2"><i class="fa fa-check"></i><b>6.2</b> About ggplot2</a></li>
<li class="chapter" data-level="6.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#plotting-bivariate-data-continuous-y-by-continuous-x"><i class="fa fa-check"></i><b>6.3</b> Plotting bivariate data: continuous Y by continuous X</a><ul>
<li class="chapter" data-level="6.3.1" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#adding-layers"><i class="fa fa-check"></i><b>6.3.1</b> Adding layers</a></li>
<li class="chapter" data-level="6.3.2" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#faceting"><i class="fa fa-check"></i><b>6.3.2</b> Faceting</a></li>
<li class="chapter" data-level="6.3.3" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#saving-plots"><i class="fa fa-check"></i><b>6.3.3</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#plotting-bivariate-data-continuous-y-by-categorical-x"><i class="fa fa-check"></i><b>6.4</b> Plotting bivariate data: continuous Y by categorical X</a></li>
<li class="chapter" data-level="6.5" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#plotting-univariate-continuous-data"><i class="fa fa-check"></i><b>6.5</b> Plotting univariate continuous data</a></li>
<li class="chapter" data-level="6.6" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#publication-ready-plots-themes"><i class="fa fa-check"></i><b>6.6</b> Publication-ready plots &amp; themes</a></li>
<li class="chapter" data-level="6.7" data-path="data-visualization-with-ggplot2.html"><a href="data-visualization-with-ggplot2.html#homework-1"><i class="fa fa-check"></i><b>6.7</b> Homework</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="reproducible-research.html"><a href="reproducible-research.html"><i class="fa fa-check"></i><b>7</b> Reproducible Research</a><ul>
<li class="chapter" data-level="7.1" data-path="reproducible-research.html"><a href="reproducible-research.html#before-class"><i class="fa fa-check"></i><b>7.1</b> Before class</a></li>
<li class="chapter" data-level="7.2" data-path="reproducible-research.html"><a href="reproducible-research.html#who-cares-about-reproducible-research"><i class="fa fa-check"></i><b>7.2</b> Who cares about reproducible research?</a><ul>
<li class="chapter" data-level="7.2.1" data-path="reproducible-research.html"><a href="reproducible-research.html#reproducibility-is-hard"><i class="fa fa-check"></i><b>7.2.1</b> Reproducibility is hard!</a></li>
<li class="chapter" data-level="7.2.2" data-path="reproducible-research.html"><a href="reproducible-research.html#whats-in-it-for-you"><i class="fa fa-check"></i><b>7.2.2</b> What’s in it for <em>you?</em></a></li>
<li class="chapter" data-level="7.2.3" data-path="reproducible-research.html"><a href="reproducible-research.html#some-recommendations-for-reproducible-research"><i class="fa fa-check"></i><b>7.2.3</b> Some recommendations for reproducible research</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="reproducible-research.html"><a href="reproducible-research.html#rmarkdown"><i class="fa fa-check"></i><b>7.3</b> RMarkdown</a><ul>
<li class="chapter" data-level="7.3.1" data-path="reproducible-research.html"><a href="reproducible-research.html#markdown"><i class="fa fa-check"></i><b>7.3.1</b> Markdown</a></li>
<li class="chapter" data-level="7.3.2" data-path="reproducible-research.html"><a href="reproducible-research.html#rmarkdown-workflow"><i class="fa fa-check"></i><b>7.3.2</b> RMarkdown workflow</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="reproducible-research.html"><a href="reproducible-research.html#authoring-rmarkdown-documents"><i class="fa fa-check"></i><b>7.4</b> Authoring RMarkdown documents</a><ul>
<li class="chapter" data-level="7.4.1" data-path="reproducible-research.html"><a href="reproducible-research.html#from-scratch"><i class="fa fa-check"></i><b>7.4.1</b> From scratch</a></li>
<li class="chapter" data-level="7.4.2" data-path="reproducible-research.html"><a href="reproducible-research.html#from-a-template-with-yaml-metadata"><i class="fa fa-check"></i><b>7.4.2</b> From a template with YAML metadata</a></li>
<li class="chapter" data-level="7.4.3" data-path="reproducible-research.html"><a href="reproducible-research.html#chunk-options"><i class="fa fa-check"></i><b>7.4.3</b> Chunk options</a></li>
<li class="chapter" data-level="7.4.4" data-path="reproducible-research.html"><a href="reproducible-research.html#tables"><i class="fa fa-check"></i><b>7.4.4</b> Tables</a></li>
<li class="chapter" data-level="7.4.5" data-path="reproducible-research.html"><a href="reproducible-research.html#changing-output-formats"><i class="fa fa-check"></i><b>7.4.5</b> Changing output formats</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="reproducible-research.html"><a href="reproducible-research.html#distributing-analyses-rpubs"><i class="fa fa-check"></i><b>7.5</b> Distributing Analyses: Rpubs</a></li>
<li class="chapter" data-level="7.6" data-path="reproducible-research.html"><a href="reproducible-research.html#further-resources"><i class="fa fa-check"></i><b>7.6</b> Further resources</a></li>
<li class="chapter" data-level="7.7" data-path="reproducible-research.html"><a href="reproducible-research.html#homework-2"><i class="fa fa-check"></i><b>7.7</b> Homework</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="essential-statistics.html"><a href="essential-statistics.html"><i class="fa fa-check"></i><b>8</b> Essential Statistics</a><ul>
<li class="chapter" data-level="8.1" data-path="essential-statistics.html"><a href="essential-statistics.html#our-data-nhanes"><i class="fa fa-check"></i><b>8.1</b> Our data: NHANES</a><ul>
<li class="chapter" data-level="8.1.1" data-path="essential-statistics.html"><a href="essential-statistics.html#about-nhanes"><i class="fa fa-check"></i><b>8.1.1</b> About NHANES</a></li>
<li class="chapter" data-level="8.1.2" data-path="essential-statistics.html"><a href="essential-statistics.html#import-inspect"><i class="fa fa-check"></i><b>8.1.2</b> Import &amp; inspect</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="essential-statistics.html"><a href="essential-statistics.html#descriptive-statistics"><i class="fa fa-check"></i><b>8.2</b> Descriptive statistics</a><ul>
<li class="chapter" data-level="8.2.1" data-path="essential-statistics.html"><a href="essential-statistics.html#missing-data"><i class="fa fa-check"></i><b>8.2.1</b> Missing data</a></li>
<li class="chapter" data-level="8.2.2" data-path="essential-statistics.html"><a href="essential-statistics.html#eda"><i class="fa fa-check"></i><b>8.2.2</b> EDA</a></li>
<li class="chapter" data-level="8.2.3" data-path="essential-statistics.html"><a href="essential-statistics.html#exercise-set-1"><i class="fa fa-check"></i><b>8.2.3</b> Exercise set 1</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="essential-statistics.html"><a href="essential-statistics.html#continuous-variables"><i class="fa fa-check"></i><b>8.3</b> Continuous variables</a><ul>
<li class="chapter" data-level="8.3.1" data-path="essential-statistics.html"><a href="essential-statistics.html#t-tests"><i class="fa fa-check"></i><b>8.3.1</b> T-tests</a></li>
<li class="chapter" data-level="8.3.2" data-path="essential-statistics.html"><a href="essential-statistics.html#wilcoxon-test"><i class="fa fa-check"></i><b>8.3.2</b> Wilcoxon test</a></li>
<li class="chapter" data-level="8.3.3" data-path="essential-statistics.html"><a href="essential-statistics.html#linear-models"><i class="fa fa-check"></i><b>8.3.3</b> Linear models</a></li>
<li class="chapter" data-level="8.3.4" data-path="essential-statistics.html"><a href="essential-statistics.html#anova"><i class="fa fa-check"></i><b>8.3.4</b> ANOVA</a></li>
<li class="chapter" data-level="8.3.5" data-path="essential-statistics.html"><a href="essential-statistics.html#linear-regression"><i class="fa fa-check"></i><b>8.3.5</b> Linear regression</a></li>
<li class="chapter" data-level="8.3.6" data-path="essential-statistics.html"><a href="essential-statistics.html#multiple-regression"><i class="fa fa-check"></i><b>8.3.6</b> Multiple regression</a></li>
<li class="chapter" data-level="8.3.7" data-path="essential-statistics.html"><a href="essential-statistics.html#exercise-set-2"><i class="fa fa-check"></i><b>8.3.7</b> Exercise set 2</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="essential-statistics.html"><a href="essential-statistics.html#discrete-variables"><i class="fa fa-check"></i><b>8.4</b> Discrete variables</a><ul>
<li class="chapter" data-level="8.4.1" data-path="essential-statistics.html"><a href="essential-statistics.html#contingency-tables"><i class="fa fa-check"></i><b>8.4.1</b> Contingency tables</a></li>
<li class="chapter" data-level="8.4.2" data-path="essential-statistics.html"><a href="essential-statistics.html#logistic-regression"><i class="fa fa-check"></i><b>8.4.2</b> Logistic regression</a></li>
<li class="chapter" data-level="8.4.3" data-path="essential-statistics.html"><a href="essential-statistics.html#exercise-set-3"><i class="fa fa-check"></i><b>8.4.3</b> Exercise set 3</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="essential-statistics.html"><a href="essential-statistics.html#power-sample-size"><i class="fa fa-check"></i><b>8.5</b> Power &amp; sample size</a><ul>
<li class="chapter" data-level="8.5.1" data-path="essential-statistics.html"><a href="essential-statistics.html#t-test-powern"><i class="fa fa-check"></i><b>8.5.1</b> T-test power/N</a></li>
<li class="chapter" data-level="8.5.2" data-path="essential-statistics.html"><a href="essential-statistics.html#proportions-powern"><i class="fa fa-check"></i><b>8.5.2</b> Proportions power/N</a></li>
<li class="chapter" data-level="8.5.3" data-path="essential-statistics.html"><a href="essential-statistics.html#exercise-set-4"><i class="fa fa-check"></i><b>8.5.3</b> Exercise set 4</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="essential-statistics.html"><a href="essential-statistics.html#tidying-models"><i class="fa fa-check"></i><b>8.6</b> Tidying models</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="surival-analysis.html"><a href="surival-analysis.html"><i class="fa fa-check"></i><b>9</b> Surival Analysis</a><ul>
<li class="chapter" data-level="9.1" data-path="surival-analysis.html"><a href="surival-analysis.html#background"><i class="fa fa-check"></i><b>9.1</b> Background</a><ul>
<li class="chapter" data-level="9.1.1" data-path="surival-analysis.html"><a href="surival-analysis.html#definitions"><i class="fa fa-check"></i><b>9.1.1</b> Definitions</a></li>
<li class="chapter" data-level="9.1.2" data-path="surival-analysis.html"><a href="surival-analysis.html#cox-ph-model"><i class="fa fa-check"></i><b>9.1.2</b> Cox PH Model</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="surival-analysis.html"><a href="surival-analysis.html#survival-analysis-in-r"><i class="fa fa-check"></i><b>9.2</b> Survival analysis in R</a><ul>
<li class="chapter" data-level="9.2.1" data-path="surival-analysis.html"><a href="surival-analysis.html#getting-started"><i class="fa fa-check"></i><b>9.2.1</b> Getting started</a></li>
<li class="chapter" data-level="9.2.2" data-path="surival-analysis.html"><a href="surival-analysis.html#survival-curves"><i class="fa fa-check"></i><b>9.2.2</b> Survival Curves</a></li>
<li class="chapter" data-level="9.2.3" data-path="surival-analysis.html"><a href="surival-analysis.html#kaplan-meier-plots"><i class="fa fa-check"></i><b>9.2.3</b> Kaplan-Meier Plots</a></li>
<li class="chapter" data-level="9.2.4" data-path="surival-analysis.html"><a href="surival-analysis.html#exercise-set-1-1"><i class="fa fa-check"></i><b>9.2.4</b> Exercise set 1</a></li>
<li class="chapter" data-level="9.2.5" data-path="surival-analysis.html"><a href="surival-analysis.html#cox-regression"><i class="fa fa-check"></i><b>9.2.5</b> Cox Regression</a></li>
<li class="chapter" data-level="9.2.6" data-path="surival-analysis.html"><a href="surival-analysis.html#exercise-set-2-1"><i class="fa fa-check"></i><b>9.2.6</b> Exercise set 2</a></li>
<li class="chapter" data-level="9.2.7" data-path="surival-analysis.html"><a href="surival-analysis.html#categorizing-for-km-plots"><i class="fa fa-check"></i><b>9.2.7</b> Categorizing for KM plots</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="surival-analysis.html"><a href="surival-analysis.html#tcga"><i class="fa fa-check"></i><b>9.3</b> TCGA</a><ul>
<li class="chapter" data-level="9.3.1" data-path="surival-analysis.html"><a href="surival-analysis.html#rtcga"><i class="fa fa-check"></i><b>9.3.1</b> RTCGA</a></li>
<li class="chapter" data-level="9.3.2" data-path="surival-analysis.html"><a href="surival-analysis.html#exercise-set-3-1"><i class="fa fa-check"></i><b>9.3.2</b> Exercise set 3</a></li>
<li class="chapter" data-level="9.3.3" data-path="surival-analysis.html"><a href="surival-analysis.html#other-tcga-resources"><i class="fa fa-check"></i><b>9.3.3</b> Other TCGA Resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html"><i class="fa fa-check"></i><b>10</b> RNA-Seq Analysis</a><ul>
<li class="chapter" data-level="10.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#review-3"><i class="fa fa-check"></i><b>10.1</b> Review</a><ul>
<li class="chapter" data-level="10.1.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#prerequsite-skills"><i class="fa fa-check"></i><b>10.1.1</b> Prerequsite skills</a></li>
<li class="chapter" data-level="10.1.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#data-needed-1"><i class="fa fa-check"></i><b>10.1.2</b> Data needed</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#background-1"><i class="fa fa-check"></i><b>10.2</b> Background</a><ul>
<li class="chapter" data-level="10.2.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#the-biology"><i class="fa fa-check"></i><b>10.2.1</b> The biology</a></li>
<li class="chapter" data-level="10.2.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#data-pre-processing"><i class="fa fa-check"></i><b>10.2.2</b> Data pre-processing</a></li>
<li class="chapter" data-level="10.2.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#data-structure"><i class="fa fa-check"></i><b>10.2.3</b> Data structure</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#import-data"><i class="fa fa-check"></i><b>10.3</b> Import data</a></li>
<li class="chapter" data-level="10.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#poor-mans-dge"><i class="fa fa-check"></i><b>10.4</b> Poor man’s DGE</a><ul>
<li class="chapter" data-level="10.4.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#exercise-1"><i class="fa fa-check"></i><b>10.4.1</b> Exercise 1</a></li>
<li class="chapter" data-level="10.4.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#exercise-2"><i class="fa fa-check"></i><b>10.4.2</b> Exercise 2</a></li>
<li class="chapter" data-level="10.4.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#exercise-3"><i class="fa fa-check"></i><b>10.4.3</b> Exercise 3</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#deseq2-analysis"><i class="fa fa-check"></i><b>10.5</b> DESeq2 analysis</a><ul>
<li class="chapter" data-level="10.5.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#deseq2-package"><i class="fa fa-check"></i><b>10.5.1</b> DESeq2 package</a></li>
<li class="chapter" data-level="10.5.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#importing-data"><i class="fa fa-check"></i><b>10.5.2</b> Importing data</a></li>
<li class="chapter" data-level="10.5.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#deseq-pipeline"><i class="fa fa-check"></i><b>10.5.3</b> DESeq pipeline</a></li>
<li class="chapter" data-level="10.5.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#getting-results"><i class="fa fa-check"></i><b>10.5.4</b> Getting results</a></li>
<li class="chapter" data-level="10.5.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#exercise-4"><i class="fa fa-check"></i><b>10.5.5</b> Exercise 4</a></li>
</ul></li>
<li class="chapter" data-level="10.6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#data-visualization"><i class="fa fa-check"></i><b>10.6</b> Data Visualization</a><ul>
<li class="chapter" data-level="10.6.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#plotting-counts"><i class="fa fa-check"></i><b>10.6.1</b> Plotting counts</a></li>
<li class="chapter" data-level="10.6.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#ma-volcano-plots"><i class="fa fa-check"></i><b>10.6.2</b> MA &amp; Volcano plots</a></li>
<li class="chapter" data-level="10.6.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#exercise-5"><i class="fa fa-check"></i><b>10.6.3</b> Exercise 5</a></li>
<li class="chapter" data-level="10.6.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#transformation"><i class="fa fa-check"></i><b>10.6.4</b> Transformation</a></li>
<li class="chapter" data-level="10.6.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#pca"><i class="fa fa-check"></i><b>10.6.5</b> PCA</a></li>
<li class="chapter" data-level="10.6.6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#bonus-heatmaps"><i class="fa fa-check"></i><b>10.6.6</b> Bonus: Heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="10.7" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#record-sessioninfo"><i class="fa fa-check"></i><b>10.7</b> Record <code>sessionInfo()</code></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Biomedical Data Science</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="essential-statistics" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> Essential Statistics</h1>
<p>This lesson will provide hands-on instruction and exercises covering basic statistical analysis in R. This will cover descriptive statistics, <em>t</em>-tests, linear models, chi-square, clustering, dimensionality reduction, and resampling strategies. We will also cover methods for “tidying” model results for downstream visualization and summarization.</p>
<p><strong>Prerequisites: <a href="r-basics.html">Familiarity with R</a> is <em>required</em></strong> (including working with <a href="r-dataframes.html">data frames</a>, installing/using packages, importing data, and saving results); familiarity with <a href="r-dplyr-yeast.html">dplyr</a> and <a href="r-viz-gapminder.html">ggplot2</a> packages is highly recommended.</p>
<p><strong>You must <a href="setup.html#R">complete the basic R setup here</a> <em>prior to class</em>.</strong> This includes installing R, RStudio, and the required packages. Please <a href="people.html">contact one of the instructors</a> <em>prior to class</em> if you are having difficulty with any of the setup. Please bring your laptop and charger cable to class.</p>
<p><strong>Handouts</strong>: Download and print out these handouts and bring them to class:</p>
<ul>
<li><a href="handouts/r-stats-cheatsheet.pdf">Cheat sheet</a></li>
<li><a href="handouts/r-stats-exercises.pdf">Exercises handout</a></li>
</ul>
<p><strong>Slides</strong>: <a href="slides/r-stats.html">click here</a>.</p>
<div id="our-data-nhanes" class="section level2">
<h2><span class="header-section-number">8.1</span> Our data: NHANES</h2>
<div id="about-nhanes" class="section level3">
<h3><span class="header-section-number">8.1.1</span> About NHANES</h3>
<p>The data we’re going to work with comes from the National Health and Nutrition Examination Survey (NHANES) program at the CDC. You can read a lot more about NHANES on the <a href="http://www.cdc.gov/nchs/nhanes/">CDC’s website</a> or <a href="https://en.wikipedia.org/wiki/National_Health_and_Nutrition_Examination_Survey">Wikipedia</a>. NHANES is a research program designed to assess the health and nutritional status of adults and children in the United States. The survey is one of the only to combine both survey questions and physical examinations. It began in the 1960s and since 1999 examines a nationally representative sample of about 5,000 people each year. The NHANES interview includes demographic, socioeconomic, dietary, and health-related questions. The physical exam includes medical, dental, and physiological measurements, as well as several standard laboratory tests. NHANES is used to determine the prevalence of major diseases and risk factors for those diseases. NHANES data are also the basis for national standards for measurements like height, weight, and blood pressure. Data from this survey is used in epidemiology studies and health sciences research, which help develop public health policy, direct and design health programs and services, and expand the health knowledge for the Nation.</p>
<p>We are using a small slice of this data. We’re only using a handful of variables from the 2011-2012 survey years on about 5,000 individuals. The CDC uses a <a href="http://www.cdc.gov/nchs/data/series/sr_02/sr02_162.pdf">sampling strategy</a> to purposefully oversample certain subpopulations like racial minorities. Naive analysis of the original NHANES data can lead to mistaken conclusions because the percentages of people from each racial group in the data are different from general population. The 5,000 individuals here are resampled from the larger NHANES study population to undo these oversampling effects, so you can treat this as if it were a simple random sample from the American population.</p>
<p>You can download the data at <a href="data.html">the link above</a>. The file is called <a href="data/nhanes.csv"><strong>nhanes.csv</strong></a>. There’s also a <a href="data/nhanes_dd.csv">data dictionary (filename <strong>nhanes_dd.csv</strong>)</a> that lists and describes each variable in our NHANES dataset. This table is copied below.</p>
</div>
<div id="import-inspect" class="section level3">
<h3><span class="header-section-number">8.1.2</span> Import &amp; inspect</h3>
<p>There are some built-in functions for reading in data in text files. These functions are <em>read-dot-something</em> – for example, <code>read.csv()</code> reads in comma-delimited text data; <code>read.delim()</code> reads in tab-delimited text, etc. Normally I would use instead use the <em>read-underscore-something</em> functions like <code>read_csv()</code> or <code>read_tsv()</code> from the <a href="https://github.com/hadley/readr">readr</a> package. Compared to the base functions, they’re <em>much</em> faster, they’re good at guessing the types of data in the columns, and they have more sensible defaults than the equivalent base functions do. If you have the <a href="https://cran.r-project.org/web/packages/dplyr/index.html">dplyr</a> package loaded as well, and you read in the data with readr, the data will display nicely.</p>
<p>We’re going to be working with <em>factors</em> throughout this lesson. Factors are R’s way of representing categorical variables that can take on a <em>discrete</em>, distinct set of values, e.g., “Yes/No”, “Mutant/WT”, “Case/Control”, etc. The readr package’s <code>read_*</code> functions don’t allow you to convert strings like “Yes/No” to factors automatically. But the base <code>read.*</code> functions do this automatically (not always the desired behavior).</p>
<p>We’ll want to use the <strong>dplyr</strong> package throughout, so let’s go ahead and load it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr)</code></pre></div>
<p>If you see a warning that looks like this: <code>Error in library(dplyr) : there is no package called 'dplyr'</code>, then you don’t have the package installed correctly. See the <a href="setup.html">setup page</a>.</p>
<p>Now, let’s actually load the data. You can get help for the import function with <code>?read.csv</code>. When we load data we assign it to a variable just like any other, and we can choose a name for that data. Since we’re going to be referring to this data a lot, let’s give it a short easy name to type. I’m going to call it <code>nh</code>. Once we’ve loaded it we can type the name of the object itself (<code>nh</code>) to see it printed to the screen. However, <strong>dplyr</strong> can convert your “regular” data frame to a “tibble,” which prints more nicely.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nh &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file=</span><span class="st">&quot;data/nhanes.csv&quot;</span>)
nh &lt;-<span class="st"> </span><span class="kw">tbl_df</span>(nh)
nh</code></pre></div>
<p>Alternatively, you <em>could</em> still read in the data with <strong>readr</strong>’s <code>read_csv()</code>, but you’ll want to convert the character variables to factors for the remainder of the lesson.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(readr)
nh &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;data/nhanes.csv&quot;</span>) %&gt;%<span class="st"> </span><span class="kw">mutate_if</span>(is.character, factor)
nh</code></pre></div>
<p>Take a look at that output. The nice thing about loading dplyr and wrapping the object with <code>tbl_df()</code> is that data frames are displayed in a much more friendly way. This dataset has 5,000 rows and 32 columns. When you import/convert data this way and try to display the object in the console, instead of trying to display all 5,000 rows, you’ll only see about 10 by default. Also, if you have so many columns that the data would wrap off the edge of your screen, those columns will not be displayed, but you’ll see at the bottom of the output which, if any, columns were hidden from view. If you want to see the whole dataset, there are two ways to do this. First, you can click on the name of the data.frame in the <strong>Environment</strong> panel in RStudio. Or you could use the <code>View()</code> function (<em>with a capital V</em>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">View</span>(nh)</code></pre></div>
<p>Recall several built-in functions that are useful for working with data frames.</p>
<ul>
<li>Content:
<ul>
<li><code>head()</code>: shows the first few rows</li>
<li><code>tail()</code>: shows the last few rows</li>
</ul></li>
<li>Size:
<ul>
<li><code>dim()</code>: returns a 2-element vector with the number of rows in the first element, and the number of columns as the second element (the dimensions of the object)</li>
<li><code>nrow()</code>: returns the number of rows</li>
<li><code>ncol()</code>: returns the number of columns</li>
</ul></li>
<li>Summary:
<ul>
<li><code>colnames()</code> (or just <code>names()</code>): returns the column names</li>
<li><code>glimpse()</code> (from <strong>dplyr</strong>): Returns a glimpse of your data, telling you the structure of the dataset and information about the class, length and content of each column</li>
</ul></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(nh)
<span class="kw">tail</span>(nh)
<span class="kw">dim</span>(nh)
<span class="kw">names</span>(nh)
<span class="kw">glimpse</span>(nh)</code></pre></div>
</div>
</div>
<div id="descriptive-statistics" class="section level2">
<h2><span class="header-section-number">8.2</span> Descriptive statistics</h2>
<p>We can access individual variables within a data frame using the <code>$</code> operator, e.g., <code>mydataframe$specificVariable</code>. Let’s print out all the <strong><code>Race</code></strong> values in the data. Let’s then see what are the <code>unique</code> values of each. Then let’s calculate the <code>mean</code>, <code>median</code>, and <code>range</code> of the <strong><code>Age</code></strong> variable.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Display all Race values</span>
nh$Race

<span class="co"># Get the unique values of Race</span>
<span class="kw">unique</span>(nh$Race)
<span class="kw">length</span>(<span class="kw">unique</span>(nh$Race))
<span class="co"># Do the same thing the dplyr way</span>
nh$Race %&gt;%<span class="st"> </span><span class="kw">unique</span>()
nh$Race %&gt;%<span class="st"> </span><span class="kw">unique</span>() %&gt;%<span class="st"> </span><span class="kw">length</span>()

<span class="co"># Age mean, median, range</span>
<span class="kw">mean</span>(nh$Age)
<span class="kw">median</span>(nh$Age)
<span class="kw">range</span>(nh$Age)</code></pre></div>
<p>The <code>summary()</code> function (note, this is different from <strong>dplyr</strong>’s <code>summarize()</code>) works differently depending on which kind of object you pass to it. If you run <code>summary()</code> on a data frame, you get some very basic summary statistics on each variable in the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(nh)</code></pre></div>
<div id="missing-data" class="section level3">
<h3><span class="header-section-number">8.2.1</span> Missing data</h3>
<p>Let’s try taking the mean of a different variable.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(nh$Income)</code></pre></div>
<p>What happened there? <code>NA</code> indicates <em>missing data</em>. Take a look at the Income variable.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Look at just the Income variable</span>
nh$Income

<span class="co"># Or view the dataset</span>
<span class="co"># View(nh)</span></code></pre></div>
<p>Notice that there are lots of missing values for Income. Trying to get the mean a bunch of observations with some missing data returns a missing value by default. This is almost universally the case with all summary statistics – a single <code>NA</code> will cause the summary to return <code>NA</code>. Now look at the help for <code>?mean</code>. Notice the <code>na.rm</code> argument. This is a logical (i.e., <code>TRUE</code> or <code>FALSE</code>) value indicating whether or not missing values should be removed prior to computing the mean. By default, it’s set to <code>FALSE</code>. Now try it again.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(nh$Income, <span class="dt">na.rm=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>The <code>is.na()</code> function tells you if a value is missing. Get the <code>sum()</code> of that vector, which adds up all the <code>TRUE</code>s to tell you how many of the values are missing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">is.na</span>(nh$Income)
<span class="kw">sum</span>(<span class="kw">is.na</span>(nh$Income))</code></pre></div>
<p>There are a few handy functions in the <a href="https://cran.r-project.org/web/packages/Tmisc/index.html"><strong>Tmisc</strong> package</a> for summarizing missingness in a data frame. You can graphically display missingness in a data frame as holes on a black canvas with <strong><code>gg_na()</code></strong> (use ggplot2 to plot <code>NA</code> values), or show a table of all the variables and the missingness level with <strong><code>propmiss()</code></strong>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Install if you don&#39;t have it already</span>
<span class="co"># install.packages(&quot;Tmisc&quot;)</span>

<span class="co"># Load Tmisc</span>
<span class="kw">library</span>(Tmisc)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">gg_na</span>(nh)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">propmiss</span>(nh)</code></pre></div>
<p>Now, let’s talk about exploratory data analysis (EDA).</p>
</div>
<div id="eda" class="section level3">
<h3><span class="header-section-number">8.2.2</span> EDA</h3>
<p>It’s always worth examining your data visually before you start any statistical analysis or hypothesis testing. We could spend an entire day on <strong><a href="https://en.wikipedia.org/wiki/Exploratory_data_analysis">exploratory data analysis</a></strong>. The <a href="r-viz-gapminder.html">data visualization lesson</a> covers this in much broader detail. Here we’ll just mention a few of the big ones: <strong>histograms</strong> and <strong>scatterplots</strong>.</p>
<div id="histograms" class="section level4">
<h4><span class="header-section-number">8.2.2.1</span> Histograms</h4>
<p>We can learn a lot from the data just looking at the value distributions of particular variables. Let’s make some histograms with ggplot2. Looking at BMI shows a few extreme outliers. Looking at weight initially shows us that the units are probably in kg. Replotting that in lbs with more bins shows a clear bimodal distribution. Are there kids in this data? The age distribution shows us the answer is <em>yes</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">ggplot</span>(nh, <span class="kw">aes</span>(BMI)) +<span class="st"> </span><span class="kw">geom_histogram</span>()
<span class="kw">ggplot</span>(nh, <span class="kw">aes</span>(Weight)) +<span class="st"> </span><span class="kw">geom_histogram</span>()
<span class="co"># In pounds, more bins</span>
<span class="kw">ggplot</span>(nh, <span class="kw">aes</span>(Weight*<span class="fl">2.2</span>)) +<span class="st"> </span><span class="kw">geom_histogram</span>(<span class="dt">bins=</span><span class="dv">80</span>)
<span class="kw">ggplot</span>(nh, <span class="kw">aes</span>(Age)) +<span class="st"> </span><span class="kw">geom_histogram</span>()</code></pre></div>
</div>
<div id="scatterplots" class="section level4">
<h4><span class="header-section-number">8.2.2.2</span> Scatterplots</h4>
<p>Let’s look at how a few different variables relate to each other. E.g., height and weight:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(nh, <span class="kw">aes</span>(Height, Weight, <span class="dt">col=</span>Gender)) +<span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p>Let’s filter out all the kids, draw trend lines using a linear model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nh %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(Age&gt;=<span class="dv">18</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(Height, Weight, <span class="dt">col=</span>Gender)) +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_point</span>() +<span class="st"> </span>
<span class="st">    </span><span class="kw">geom_smooth</span>(<span class="dt">method=</span><span class="st">&quot;lm&quot;</span>)</code></pre></div>
<p>Check out the <a href="r-viz-gapminder.html">data visualization lesson</a> for much more on this topic.</p>
</div>
</div>
<div id="exercise-set-1" class="section level3">
<h3><span class="header-section-number">8.2.3</span> Exercise set 1</h3>
<ol style="list-style-type: decimal">
<li><p>What’s the mean 60-second pulse rate for all participants in the data?</p></li>
<li><p>What’s the range of values for diastolic blood pressure in all participants? (Hint: see help for <code>min()</code>, <code>max()</code>, and <code>range()</code> functions, e.g., enter <code>?range</code> without the parentheses to get help).</p></li>
<li><p>What are the median, lower, and upper quartiles for the age of all participants? (Hint: see help for <code>median</code>, or better yet, <code>quantile</code>).</p></li>
<li><p>What’s the variance and standard deviation for income among all participants?</p></li>
</ol>
</div>
</div>
<div id="continuous-variables" class="section level2">
<h2><span class="header-section-number">8.3</span> Continuous variables</h2>
<div id="t-tests" class="section level3">
<h3><span class="header-section-number">8.3.1</span> T-tests</h3>
<p>First let’s create a new dataset from <code>nh</code> called <code>nha</code> that only has adults. To prevent us from making any mistakes downstream, let’s remove the <code>nh</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nha &lt;-<span class="st"> </span><span class="kw">filter</span>(nh, Age&gt;=<span class="dv">18</span>)
<span class="kw">rm</span>(nh)
<span class="co"># View(nha)</span></code></pre></div>
<p>Let’s do a few two-sample t-tests to test for <em>differences in means between two groups</em>. The function for a t-test is <code>t.test()</code>. See the help for <code>?t.test</code>. We’ll be using the <em>forumla</em> method. The usage is <strong><code>t.test(response~group, data=myDataFrame)</code></strong>.</p>
<ol style="list-style-type: decimal">
<li>Are there differences in age for males versus females in this dataset?</li>
<li>Does BMI differ between diabetics and non-diabetics?</li>
<li>Do single or married/cohabitating people drink more alcohol? Is this relationship significant?</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">t.test</span>(Age~Gender, <span class="dt">data=</span>nha)
<span class="kw">t.test</span>(BMI~Diabetes, <span class="dt">data=</span>nha)
<span class="kw">t.test</span>(AlcoholYear~RelationshipStatus, <span class="dt">data=</span>nha)</code></pre></div>
<p>See the heading, <em>Welch Two Sample t-test</em>, and notice that the degrees of freedom might not be what we expected based on our sample size. Now look at the help for <code>?t.test</code> again, and look at the <code>var.equal</code> argument, which is by default set to <code>FALSE</code>. One of the assumptions of the t-test is <a href="https://en.wikipedia.org/wiki/Homoscedasticity">homoscedasticity</a>, or homogeneity of variance. This assumes that the variance in the outcome (e.g., BMI) is identical across both levels of the predictor (diabetic vs non-diabetic). Since this is rarely the case, the t-test defaults to using the <a href="https://en.wikipedia.org/wiki/Welch%27s_t-test">Welch correction</a>, which is a more reliable version of the t-test when the homoscedasticity assumption is violated.</p>
</div>
<div id="wilcoxon-test" class="section level3">
<h3><span class="header-section-number">8.3.2</span> Wilcoxon test</h3>
<p>Another assumption of the t-test is that data is normally distributed. Looking at the histogram for AlcoholYear shows that this data clearly isn’t.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(nha, <span class="kw">aes</span>(AlcoholYear)) +<span class="st"> </span><span class="kw">geom_histogram</span>()</code></pre></div>
<p>The <a href="https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test">Wilcoxon rank-sum test (a.k.a. Mann-Whitney <em>U</em> test)</a> is a nonparametric test of differences in mean that does not require normally distributed data. When data is perfectly normal, the t-test is uniformly more powerful. But when this assumption is violated, the t-test is unreliable. This test is called in a similar way as the t-test.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">wilcox.test</span>(AlcoholYear~RelationshipStatus, <span class="dt">data=</span>nha)</code></pre></div>
<p>The results are still significant, but much less than the p-value reported for the (incorrect) t-test above.</p>
</div>
<div id="linear-models" class="section level3">
<h3><span class="header-section-number">8.3.3</span> Linear models</h3>
<blockquote>
<p>Analysis of variance and linear modeling are complex topics that deserve an entire semester dedicated to theory, design, and interpretation. A very good resource is <a href="https://www.amazon.com/Introduction-Statistical-Learning-Applications-Statistics/dp/1461471370/ref=sr_1_1?ie=UTF8&amp;qid=1473087847&amp;sr=8-1&amp;keywords=introduction+statistical+learning&amp;tag=gettgenedone-20"><em>An Introduction to Statistical Learning: with Applications in R</em></a> by Gareth James, Daniela Witten, Trevor Hastie and Robert Tibshirani. The <a href="http://www-bcf.usc.edu/~gareth/ISL/ISLR%20Sixth%20Printing.pdf">PDF</a> of the book and all the R code used throughout are <a href="http://www-bcf.usc.edu/~gareth/ISL/">available <strong>free</strong> on the author’s website</a>. What follows is a necessary over-simplification with more focus on implementation, and less on theory and design.</p>
</blockquote>
<p>Where t-tests and their nonparametric substitutes are used for assessing the differences in means between two groups, ANOVA is used to assess the significance of differences in means between multiple groups. In fact, a t-test is just a specific case of ANOVA when you only have two groups. And both t-tests and ANOVA are just specific cases of linear regression, where you’re trying to fit a model describing how a continuous outcome (e.g., BMI) changes with some predictor variable (e.g., diabetic status, race, age, etc.). The distinction is largely semantic – with a linear model you’re asking, “do levels of a categorical variable affect the response?” where with ANOVA or t-tests you’re asking, “does the mean response differ between levels of a categorical variable?”</p>
<p>Let’s examine the relationship between BMI and relationship status (<code>RelationshipStatus</code> was derived from <code>MaritalStatus</code>, coded as <em>Committed</em> if MaritalStatus is Married or LivePartner, and <em>Single</em> otherwise). Let’s first do this with a t-test, and for now, let’s assume that the variances between groups <em>are</em> equal.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">t.test</span>(BMI~RelationshipStatus, <span class="dt">data=</span>nha, <span class="dt">var.equal=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>It looks like single people have a very slightly higher BMI than those in a committed relationship, but the magnitude of the difference is trivial, and the difference is not significant. Now, let’s do the same test in a linear modeling framework. First, let’s create the fitted model and store it in an object called <code>fit</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">lm</span>(BMI~RelationshipStatus, <span class="dt">data=</span>nha)</code></pre></div>
<p>You can display the object itself, but that isn’t too interesting. You can get the more familiar ANOVA table by calling the <code>anova()</code> function on the <code>fit</code> object. More generally, the <code>summary()</code> function on a linear model object will tell you much more.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit
<span class="kw">anova</span>(fit)
<span class="kw">summary</span>(fit)</code></pre></div>
<p>Go back and re-run the t-test assuming equal variances as we did before. Now notice a few things:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">t.test</span>(BMI~RelationshipStatus, <span class="dt">data=</span>nha, <span class="dt">var.equal=</span><span class="ot">TRUE</span>)</code></pre></div>
<ol style="list-style-type: decimal">
<li>The p-values from all three tests (t-test, ANOVA, and linear regression) are all identical (p=0.1256). This is because they’re all identical: a t-test is a specific case of ANOVA, which is a specific case of linear regression. There may be some rounding error, but we’ll talk about extracting the exact values from a model object later on.</li>
<li>The test statistics are all related. The <em>t</em> statistic from the t-test is <strong>1.532</strong>, which is the same as the t-statistic from the linear regression. If you square that, you get <strong>2.347</strong>, the <em>F</em> statistic from the ANOVA.</li>
<li>The <code>t.test()</code> output shows you the means for the two groups, Committed and Single. Just displaying the <code>fit</code> object itself or running <code>summary(fit)</code> shows you the coefficients for a linear model. Here, the model assumes the “baseline” RelationshipStatus level is <em>Committed</em>, and that the <em>intercept</em> in a regression model (e.g., <span class="math inline">\(\beta_{0}\)</span> in the model <span class="math inline">\(Y = \beta_{0} + \beta_{1}X\)</span>) is the mean of the baseline group. Being <em>Single</em> results in an increase in BMI of 0.3413. This is the <span class="math inline">\(\beta_{1}\)</span> coefficient in the model. You can easily change the ordering of the levels. See the help for <code>?factor</code>, and check out the new <a href="https://hadley.github.io/forcats/"><strong>forcats</strong> package</a>, which provides tools <strong>for</strong> manipulating <strong>cat</strong>egorical variables.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># P-value computed on a t-statistic with 3552 degrees of freedom</span>
<span class="co"># (multiply times 2 because t-test is assuming two-tailed)</span>
<span class="dv">2</span>*(<span class="dv">1</span>-<span class="kw">pt</span>(<span class="fl">1.532</span>, <span class="dt">df=</span><span class="dv">3552</span>))
<span class="co"># P-value computed on an F-test with 1 and 3552 degrees of freedom</span>
<span class="dv">1</span>-<span class="kw">pf</span>(<span class="fl">2.347</span>, <span class="dt">df1=</span><span class="dv">1</span>, <span class="dt">df2=</span><span class="dv">3552</span>)</code></pre></div>
</div>
<div id="anova" class="section level3">
<h3><span class="header-section-number">8.3.4</span> ANOVA</h3>
<p>Recap: t-tests are for assessing the differences in means between <em>two</em> groups. A t-test is a specific case of ANOVA, which is a specific case of a linear model. Let’s run ANOVA, but this time looking for differences in means between more than two groups.</p>
<p>Let’s look at the relationship between smoking status (Never, Former, or Current), and BMI.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">lm</span>(BMI~SmokingStatus, <span class="dt">data=</span>nha)
<span class="kw">anova</span>(fit)
<span class="kw">summary</span>(fit)</code></pre></div>
<p>The F-test on the ANOVA table tells us that there <em>is</em> a significant difference in means between current, former, and never smokers (p=<span class="math inline">\(4.54 \times 10^{-8}\)</span>). However, the linear model output might not have been what we wanted. Because the default handling of categorical variables is to treat the alphabetical first level as the baseline, “Current” smokers are treated as baseline, and this mean becomes the intercept, and the coefficients on “Former” and “Never” describe how those groups’ means differ from current smokers. What if we wanted “Never” smokers to be the baseline, followed by Former, then Current? Have a look at <code>?factor</code> to relevel the factor levels.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Re-level the SmokingStatus variable</span>
nha$SmokingStatus &lt;-<span class="st"> </span><span class="kw">factor</span>(nha$SmokingStatus, 
                            <span class="dt">levels=</span><span class="kw">c</span>(<span class="st">&quot;Never&quot;</span>, <span class="st">&quot;Former&quot;</span>, <span class="st">&quot;Current&quot;</span>))

<span class="co"># Re-fit the model</span>
fit &lt;-<span class="st"> </span><span class="kw">lm</span>(BMI~SmokingStatus, <span class="dt">data=</span>nha)

<span class="co"># Show the ANOVA table</span>
<span class="kw">anova</span>(fit)

<span class="co"># Print the full model statistics</span>
<span class="kw">summary</span>(fit)</code></pre></div>
<p>Notice that the p-value on the ANOVA/regression didn’t change, but the coefficients did. <em>Never</em> smokers are now treated as baseline. The intercept coefficient (28.856) is now the mean for <em>Never</em> smokers. The <code>SmokingStatusFormer</code> coefficient of .309 shows the apparent increase in BMI that former smokers have when compared to never smokers, but that difference is not significant (p=.24). The <code>SmokingStatusCurrent</code> coefficient of -1.464 shows that current smokers actually have a lower BMI than never smokers, and that this decrease is highly significant.</p>
<p>Finally, you can do the typical post-hoc ANOVA procedures on the fit object. For example, the <code>TukeyHSD()</code> function will run <a href="https://en.wikipedia.org/wiki/Tukey%27s_range_test"><em>Tukey’s test</em></a> (also known as <em>Tukey’s range test</em>, the <em>Tukey method</em>, <em>Tukey’s honest significance test</em>, <em>Tukey’s HSD test</em> (honest significant difference), or the <em>Tukey-Kramer method</em>). Tukey’s test computes all pairwise mean difference calculation, comparing each group to each other group, identifying any difference between two groups that’s greater than the standard error, while controlling the type I error for all multiple comparisons. First run <code>aov()</code> (<strong>not</strong> <code>anova()</code>) on the fitted linear model object, then run <code>TukeyHSD()</code> on the resulting analysis of variance fit.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">TukeyHSD</span>(<span class="kw">aov</span>(fit))</code></pre></div>
<p>This shows that there isn’t much of a difference between former and never smokers, but that both of these differ significantly from current smokers, who have significantly lower BMI.</p>
<p>Finally, let’s visualize the differences in means between these groups. The <strong>NA</strong> category, which is omitted from the ANOVA, contains all the observations who have missing or non-recorded Smoking Status.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(nha, <span class="kw">aes</span>(SmokingStatus, BMI)) +<span class="st"> </span><span class="kw">geom_boxplot</span>() +<span class="st"> </span><span class="kw">theme_classic</span>()</code></pre></div>
</div>
<div id="linear-regression" class="section level3">
<h3><span class="header-section-number">8.3.5</span> Linear regression</h3>
<p>Linear models are mathematical representations of the process that (<em>we think</em>) gave rise to our data. The model seeks to explain the relationship between a variable of interest, our <em>Y</em>, <em>outcome</em>, <em>response</em>, or <em>dependent</em> variable, and one or more <em>X</em>, <em>predictor</em>, or <em>independent</em> variables. Previously we talked about t-tests or ANOVA in the context of a simple linear regression model with only a single predictor variable, <span class="math inline">\(X\)</span>:</p>
<p><span class="math display">\[Y = \beta_{0} +  \beta_{1}X\]</span></p>
<p>But you can have multiple predictors in a linear model that are all additive, accounting for the effects of the others:</p>
<p><span class="math display">\[Y = \beta_{0} +  \beta_{1}X_{1} + \beta_{2}X_{2} + \epsilon\]</span></p>
<ul>
<li><span class="math inline">\(Y\)</span> is the response</li>
<li><span class="math inline">\(X_{1}\)</span> and <span class="math inline">\(X_{2}\)</span> are the predictors</li>
<li><span class="math inline">\(\beta_{0}\)</span> is the intercept, and <span class="math inline">\(\beta_{1}\)</span>, <span class="math inline">\(\beta_{2}\)</span> etc are <em>coefficients</em> that describe what 1-unit changes in <span class="math inline">\(X_{1}\)</span> and <span class="math inline">\(X_{2}\)</span> do to the outcome variable <span class="math inline">\(Y\)</span>.</li>
<li><span class="math inline">\(\epsilon\)</span> is random error. Our model will not perfectly predict <span class="math inline">\(Y\)</span>. It will be off by some random amount. We assume this amount is a random draw from a Normal distribution with mean 0 and standard deviation <span class="math inline">\(\sigma\)</span>.</li>
</ul>
<p><em>Building a linear model</em> means we propose a linear model and then estimate the coefficients and the variance of the error term. Above, this means estimating <span class="math inline">\(\beta_{0}, \beta_{1}, \beta_{2}\)</span> and <span class="math inline">\(\sigma\)</span>. This is what we do in R.</p>
<p>Let’s look at the relationship between height and weight.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">lm</span>(Weight~Height, <span class="dt">data=</span>nha)
<span class="kw">summary</span>(fit)</code></pre></div>
<p>The relationship is highly significant (P&lt;<span class="math inline">\(2.2 \times 10^{-16}\)</span>). The intercept term is not very useful most of the time. Here it shows us what the value of Weight would be when Height=0, which could never happen. The Height coefficient is meaningful – each one unit increase in height results in a 0.92 increase in the corresponding unit of weight. Let’s visualize that relationship:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(nha, <span class="kw">aes</span>(<span class="dt">x=</span>Height, <span class="dt">y=</span>Weight)) +<span class="st"> </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">geom_smooth</span>(<span class="dt">method=</span><span class="st">&quot;lm&quot;</span>)</code></pre></div>
<p>By default, this is only going to show the prediction over the range of the data. This is important! You never want to try to extrapolate response variables outside of the range of your predictor(s). For example, the linear model tells us that weight is -73.7kg when height is zero. We can extend the predicted model / regression line past the lowest value of the data down to height=0. The bands on the confidence interval tell us that the model is apparently confident within the regions defined by the gray boundary. But this is silly – we would never see a height of zero, and predicting past the range of the available training data is never a good idea.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(nha, <span class="kw">aes</span>(<span class="dt">x=</span>Height, <span class="dt">y=</span>Weight)) +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_point</span>() +<span class="st"> </span>
<span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method=</span><span class="st">&quot;lm&quot;</span>, <span class="dt">fullrange=</span><span class="ot">TRUE</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">xlim</span>(<span class="dv">0</span>, <span class="ot">NA</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">&quot;Friends don&#39;t let friends extrapolate.&quot;</span>)</code></pre></div>
</div>
<div id="multiple-regression" class="section level3">
<h3><span class="header-section-number">8.3.6</span> Multiple regression</h3>
<p>Finally, let’s do a multiple linear regression analysis, where we attempt to model the effect of multiple predictor variables at once on some outcome. First, let’s look at the effect of physical activity on testosterone levels. Let’s do this with a t-test and linear regression, showing that you get the same results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">t.test</span>(Testosterone~PhysActive, <span class="dt">data=</span>nha)
<span class="kw">summary</span>(<span class="kw">lm</span>(Testosterone~PhysActive, <span class="dt">data=</span>nha))</code></pre></div>
<p>In both cases, the p-value is significant (p=0.01516), and the result suggest that increased physical activity is associated with increased testosterone levels. Does increasing your physical activity increase your testosterone levels? Or is it the other way – will increased testosterone encourage more physical activity? Or is it none of the above – is the apparent relationship between physical activity and testosterone levels only apparent because both are correlated with yet a third, unaccounted for variable? Let’s throw Age into the model as well.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(<span class="kw">lm</span>(Testosterone~PhysActive+Age, <span class="dt">data=</span>nha))</code></pre></div>
<p>This shows us that after accounting for age that the testosterone / physical activity link is no longer significant. Every 1-year increase in age results in a highly significant decrease in testosterone, and since increasing age is also likely associated with decreased physical activity, perhaps age is the confounder that makes this relationship apparent.</p>
<p>Adding other predictors can also swing things the other way. We know that men have much higher testosterone levels than females. Sex is probably the single best predictor of testosterone levels in our dataset. By not accounting for this effect, our unaccounted-for variation remains very high. By accounting for Gender, we now reduce the residual error in the model, and the physical activity effect once again becomes significant. Also notice that our model fits much better (higher R-squared), and is much more significant overall.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(<span class="kw">lm</span>(Testosterone~PhysActive+Age+Gender, <span class="dt">data=</span>nha))</code></pre></div>
<p>We’ve only looked at the <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/summary.lm.html"><code>summary()</code></a> and <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/anova.lm.html"><code>anova()</code></a> functions for extracting information from an <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/lm.html"><code>lm</code> class object</a>. There are several other accessor functions that can be used on a linear model object. Check out the help page for each one of these to learn more.</p>
<ul>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/coef.html"><code>coefficients()</code></a></li>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/predict.lm.html"><code>predict.lm()</code></a></li>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/fitted.values.html"><code>fitted.values()</code></a></li>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/residuals.html"><code>residuals()</code></a></li>
</ul>
</div>
<div id="exercise-set-2" class="section level3">
<h3><span class="header-section-number">8.3.7</span> Exercise set 2</h3>
<ol style="list-style-type: decimal">
<li>Is the average BMI different in single people versus those in a committed relationship? Perform a t-test.</li>
<li>The <code>Work</code> variable is coded “Looking” (n=159), “NotWorking” (n=1317), and “Working” (n=2230).
<ul>
<li>Fit a linear model. Assign this to an object called <code>fit</code>. What does the <code>fit</code> object tell you when you display it directly?</li>
<li>Run an <code>anova()</code> to get the ANOVA table. Is the model significant?</li>
<li>Run a Tukey test to get the pairwise contrasts. (Hint: <code>TukeyHSD()</code> on <code>aov()</code> on the fit). What do you conclude?</li>
<li>Instead of thinking of this as ANOVA, think of it as a linear model. After you’ve thought about it, get some <code>summary()</code> statistics on the fit. Do these results jive with the ANOVA model?</li>
</ul></li>
<li>Examine the relationship between HDL cholesterol levels (<code>HDLChol</code>) and whether someone has diabetes or not (<code>Diabetes</code>).
<ul>
<li>Is there a difference in means between diabetics and nondiabetics? Perform a t-test <em>without</em> a Welch correction (that is, assuming equal variances – see <code>?t.test</code> for help).</li>
<li>Do the same analysis in a linear modeling framework.</li>
<li>Does the relationship hold when adjusting for <code>Weight</code>?</li>
<li>What about when adjusting for <code>Weight</code>, <code>Age</code>, <code>Gender</code>, <code>PhysActive</code> (whether someone participates in moderate or vigorous-intensity sports, fitness or recreational activities, coded as yes/no). What is the effect of each of these explanatory variables?</li>
</ul></li>
</ol>
</div>
</div>
<div id="discrete-variables" class="section level2">
<h2><span class="header-section-number">8.4</span> Discrete variables</h2>
<p>Until now we’ve only discussed analyzing <em>continuous</em> outcomes / dependent variables. We’ve tested for differences in means between two groups with t-tests, differences among means between <em>n</em> groups with ANOVA, and more general relationships using linear regression. In all of these cases, the dependent variable, i.e., the outcome, or <span class="math inline">\(Y\)</span> variable, was <em>continuous</em>, and usually normally distributed. What if our outcome variable is <em>discrete</em>, e.g., “Yes/No”, “Mutant/WT”, “Case/Control”, etc.? Here we use a different set of procedures for assessing significant associations.</p>
<div id="contingency-tables" class="section level3">
<h3><span class="header-section-number">8.4.1</span> Contingency tables</h3>
<p>The <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/xtabs.html"><code>xtabs()</code></a> function is useful for creating contingency tables from categorical variables. Let’s create a gender by diabetes status contingency table, and assign it to an object called <strong><code>xt</code></strong>. After making the assignment, type the name of the object to view it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">xt &lt;-<span class="st"> </span><span class="kw">xtabs</span>(~Gender+Diabetes, <span class="dt">data=</span>nha)
xt</code></pre></div>
<p>There are two useful functions, <code>addmargins()</code> and <code>prop.table()</code> that add more information or manipulate how the data is displayed. By default, <code>prop.table()</code> will divide the number of observations in each cell by the total. But you may want to specify <em>which margin</em> you want to get proportions over. Let’s do this for the first (row) margin.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Add marginal totals</span>
<span class="kw">addmargins</span>(xt)

<span class="co"># Get the proportional table</span>
<span class="kw">prop.table</span>(xt)

<span class="co"># That wasn&#39;t really what we wanted. </span>
<span class="co"># Do this over the first (row) margin only.</span>
<span class="kw">prop.table</span>(xt, <span class="dt">margin=</span><span class="dv">1</span>)</code></pre></div>
<p>Looks like men have slightly higher rates of diabetes than women. But is this significant?</p>
<p>The chi-square test is used to assess the independence of these two factors. That is, if the null hypothesis that gender and diabetes are independent is true, the we would expect a proportionally equal number of diabetics across each sex. Males seem to be at slightly higher risk than females, but the difference is just short of statistically significant.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">chisq.test</span>(xt)</code></pre></div>
<p>An alternative to the chi-square test is <a href="https://en.wikipedia.org/wiki/Fisher%27s_exact_test">Fisher’s exact test</a>. Rather than relying on a critical value from a theoretical chi-square distribution, Fisher’s exact test calculates the <em>exact</em> probability of observing the contingency table as is. It’s especially useful when there are very small <em>n</em>’s in one or more of the contingency table cells. Both the chi-square and Fisher’s exact test give us p-values of approximately 0.06.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">fisher.test</span>(xt)</code></pre></div>
<p>There’s a useful plot for visualizing contingency table data called a <em>mosaic</em> plot. Call the <code>mosaicplot()</code> function on the contingency table object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mosaicplot</span>(xt, <span class="dt">main=</span><span class="ot">NA</span>)</code></pre></div>
<p>Let’s create a different contingency table, this time looking at the relationship between race and whether the person had health insurance. Display the table with marginal totals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">xt &lt;-<span class="st"> </span><span class="kw">xtabs</span>(~Race+Insured, <span class="dt">data=</span>nha)
<span class="kw">addmargins</span>(xt)</code></pre></div>
<p>Let’s do the same thing as above, this time showing the proportion of people in each race category having health insurance.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">prop.table</span>(xt, <span class="dt">margin=</span><span class="dv">1</span>)</code></pre></div>
<p>Now, let’s run a chi-square test for independence.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">chisq.test</span>(xt)</code></pre></div>
<p>The result is <em>highly</em> significant. In fact, so significant, that the display rounds off the p-value to something like <span class="math inline">\(&lt;2.2 \times 10^{-16}\)</span>. If you look at the help for <a href="https://stat.ethz.ch/R-manual/R-patched/library/stats/html/chisq.test.html"><code>?chisq.test</code></a> you’ll see that displaying the test only shows you summary information, but other components can be accessed. For example, we can easily get the actual p-value, or the expected counts under the null hypothesis of independence.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">chisq.test</span>(xt)$p.value
<span class="kw">chisq.test</span>(xt)$expected</code></pre></div>
<p>We can also make a mosaic plot similar to above:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mosaicplot</span>(xt, <span class="dt">main=</span><span class="ot">NA</span>)</code></pre></div>
<p>Finally, there’s an <em>association plot</em> that shows deviations from independence of rows and columns in a 2-dimensional contingency table. Each cell is represented by a rectangle that has (signed) height proportional to the deviation of the observed from expected counts and width proportional to the expected counts, so that the area of the box is proportional to the difference in observed and expected frequencies. The rectangles in each row are positioned relative to a baseline indicating independence. If the observed frequency of a cell is greater than the expected one, the box rises above the baseline and is shaded black; otherwise, the box falls below the baseline and is shaded red. See the help for <a href="https://stat.ethz.ch/R-manual/R-patched/library/graphics/html/assocplot.html"><code>?assocplot</code></a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">assocplot</span>(xt)</code></pre></div>
</div>
<div id="logistic-regression" class="section level3">
<h3><span class="header-section-number">8.4.2</span> Logistic regression</h3>
<p>What if we wanted to model the discrete outcome, e.g., whether someone is insured, against several other variables, similar to how we did with multiple linear regression? We can’t use linear regression because the outcome isn’t continuous – it’s binary, either <em>Yes</em> or <em>No</em>. For this we’ll use <em>logistic regression</em> to model the <em>log odds</em> of binary response. That is, instead of modeling the outcome variable, <span class="math inline">\(Y\)</span>, directly against the inputs, we’ll model the <em>log odds</em> of the outcome variable.</p>
<p>If <span class="math inline">\(p\)</span> is the probability that the individual is insured, then <span class="math inline">\(\frac{p}{1-p}\)</span> is the <a href="https://en.wikipedia.org/wiki/Odds"><em>odds</em></a> that person is insured. Then it follows that the linear model is expressed as:</p>
<p><span class="math display">\[log(\frac{p}{1-p}) = \beta_0 + \beta_1 x_1 + \cdots + \beta_k x_k\]</span></p>
<p>Where <span class="math inline">\(\beta_0\)</span> is the intercept, <span class="math inline">\(\beta_1\)</span> is the increase in the odds of the outcome for every unit increase in <span class="math inline">\(x_1\)</span>, and so on.</p>
<p>Logistic regression is a type of <em>generalized linear model</em> (GLM). We fit GLM models in R using the <code>glm()</code> function. It works like the <code>lm()</code> function except we specify which GLM to fit using the <code>family</code> argument. Logistic regression requires <code>family=binomial</code>.</p>
<p>The typical use looks like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mod &lt;-<span class="st"> </span><span class="kw">glm</span>(y ~<span class="st"> </span>x, <span class="dt">data=</span>yourdata, <span class="dt">family=</span><span class="st">&#39;binomial&#39;</span>)
<span class="kw">summary</span>(mod)</code></pre></div>
<p>Before we fit a logistic regression model let’s <em>relevel</em> the Race variable so that “White” is the baseline. We saw above that people who identify as “White” have the highest rates of being insured. When we run the logistic regression, we’ll get a separate coefficient (effect) for each level of the factor variable(s) in the model, telling you the increased odds that that level has, <em>as compared to the baseline group</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nha$Race &lt;-<span class="st"> </span><span class="kw">relevel</span>(<span class="kw">factor</span>(nha$Race), <span class="dt">ref=</span><span class="st">&quot;White&quot;</span>)</code></pre></div>
<p>Now, let’s fit a logistic regression model assessing how the odds of being insured change with different levels of race.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">glm</span>(Insured~Race, <span class="dt">data=</span>nha, <span class="dt">family=</span>binomial)
<span class="kw">summary</span>(fit)</code></pre></div>
<p>The <code>Estimate</code> column shows the log of the odds ratio – how the log odds of having health insurance changes at each level of race compared to White. The P-value for each coefficient is on the far right. This shows that <em>every</em> other race has <em>significantly less</em> rates of health insurance coverage. But, as in our multiple linear regression analysis above, are there other important variables that we’re leaving out that could alter our conclusions? Lets add a few more variables into the model to see if something else can explain the apparent Race-Insured association. Let’s add a few things likely to be involved (Age and Income), and something that’s probably irrelevant (hours slept at night).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">glm</span>(Insured~Race+Age+Income+SleepHrsNight, <span class="dt">data=</span>nha, <span class="dt">family=</span>binomial)
<span class="kw">summary</span>(fit)</code></pre></div>
<p>A few things become apparent:</p>
<ol style="list-style-type: decimal">
<li>Age and income are both highly associated with whether someone is insured. Both of these variables are highly significant (<span class="math inline">\(P&lt;2.2 \times 10^{-16}\)</span>), and the coefficient (the <code>Estimate</code> column) is positive, meaning that for each unit increase in one of these variables, the odds of being insured increases by the corresponding amount.</li>
<li>Hours slept per night is not meaningful at all.</li>
<li>After accounting for age and income, several of the race-specific differences are no longer statistically significant, but others remain so.</li>
<li>The absolute value of the test statistic (column called <code>z value</code>) can roughly be taken as an estimate of the “importance” of that variable to the overall model. So, age and income are the most important influences in this model; self-identifying as Hispanic or Mexican are also very highly important, hours slept per night isn’t important at all, and the other race categories fall somewhere in between.</li>
</ol>
<p>There is <em>much</em> more to go into with logistic regression. This lesson only scratches the surface. Missing from this lesson are things like regression diagnostics, model comparison approaches, penalization, interpretation of model coefficients, fitting interaction effects, and much more. Alan Agresti’s <a href="https://www.amazon.com/Categorical-Data-Analysis-Alan-Agresti/dp/0470463635/ref=sr_1_1?ie=UTF8&amp;qid=1473180895&amp;sr=8-1&amp;keywords=categorical+data+analysis&amp;tag=gettgenedone-20"><em>Categorical Data Analysis</em></a> has long been considered the definitive text on this topic. I also recommend Agresti’s <a href="https://www.amazon.com/Introduction-Categorical-Data-Analysis/dp/0471226181/ref=sr_1_3?ie=UTF8&amp;qid=1473180895&amp;sr=8-3&amp;keywords=categorical+data+analysis&amp;tag=gettgenedone-20"><em>Introduction to Categorical Data Analysis</em></a> (a.k.a. “Agresti lite”) for a gentler introduction.</p>
</div>
<div id="exercise-set-3" class="section level3">
<h3><span class="header-section-number">8.4.3</span> Exercise set 3</h3>
<ol style="list-style-type: decimal">
<li>What’s the relationship between diabetes and participating in rigorous physical activity or sports?
<ul>
<li>Create a contingency table with Diabetes status in rows and physical activity status in columns.</li>
<li>Display that table with margins.</li>
<li>Show the proportions of diabetics and nondiabetics, separately, who are physically active or not.</li>
<li>Is this relationship significant?</li>
<li>Create two different visualizations showing the relationship.</li>
</ul></li>
<li>Model the same association in a logistic regression framework to assess the risk of diabetes using physical activity as a predictor.
<ul>
<li>Fit a model with just physical activity as a predictor, and display a model summary.</li>
<li>Add gender to the model, and show a summary.</li>
<li>Continue adding weight and age to the model. What happens to the gender association?</li>
<li>Continue and add income to the model. What happens to the original association with physical activity?</li>
</ul></li>
</ol>
</div>
</div>
<div id="power-sample-size" class="section level2">
<h2><span class="header-section-number">8.5</span> Power &amp; sample size</h2>
<p>This is a necessarily short introduction to the concept of power and sample size calculations. <a href="https://en.wikipedia.org/wiki/Statistical_power">Statistical power</a>, also sometimes called sensitivity, is defined as the probability that your test correctly rejects the null hypothesis when the alternative hypothesis is true. That is, if there really is an effect (difference in means, association between categorical variables, etc.), how likely are you to be able to <em>detect</em> that effect at a given statistical significance level, given certain assumptions. Generally there are a few moving pieces, and if you know all but one of them, you can calculate what that last one is.</p>
<ol style="list-style-type: decimal">
<li>Power: How likely are you to detect the effect? (Usually like to see 80% or greater).</li>
<li>N: What is the sample size you have (or require)?</li>
<li>Effect size: How big is the difference in means, odds ratio, etc?</li>
</ol>
<p>If we know we want 80% power to detect a certain magnitude of difference between groups, we can calculate our required sample size. Or, if we know we can only collect 5 samples, we can calculate how likely we are to detect a particular effect. Or, we can work to solve the last one - if we want 80% power and we have 5 samples, what’s the smallest effect we can hope to detect?</p>
<p>All of these questions require certain assumptions about the data and the testing procedure. Which kind of test is being performed? What’s the true effect size (often unknown, or estimated from preliminary data), what’s the standard deviation of samples that will be collected (often unknown, or estimated from preliminary data), what’s the level of statistical significance needed (traditionally p&lt;0.05, but must consider multiple testing corrections).</p>
<div id="t-test-powern" class="section level3">
<h3><span class="header-section-number">8.5.1</span> T-test power/N</h3>
<p>The <code>power.t.test()</code> empirically estimates power or sample size of a t-test for differences in means. If we have 20 samples in each of two groups (e.g., control versus treatment), and the standard deviation for whatever we’re measuring is <strong>2.3</strong>, and we’re expecting a true difference in means between the groups of <strong>2</strong>, what’s the power to detect this effect?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">power.t.test</span>(<span class="dt">n=</span><span class="dv">20</span>, <span class="dt">delta=</span><span class="dv">2</span>, <span class="dt">sd=</span><span class="fl">2.3</span>)</code></pre></div>
<p>What’s the sample size we’d need to detect a difference of 0.8 given a standard deviation of 1.5, assuming we want 80% power?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">power.t.test</span>(<span class="dt">power=</span>.<span class="dv">80</span>, <span class="dt">delta=</span>.<span class="dv">8</span>, <span class="dt">sd=</span><span class="fl">1.5</span>)</code></pre></div>
</div>
<div id="proportions-powern" class="section level3">
<h3><span class="header-section-number">8.5.2</span> Proportions power/N</h3>
<p>What about a two-sample proportion test (e.g., chi-square test)? If we have two groups (control and treatment), and we’re measuring some outcome (e.g., infected yes/no), and we know that the proportion of infected controls is 80% but 20% in treated, what’s the power to detect this effect in 5 samples per group?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">power.prop.test</span>(<span class="dt">n=</span><span class="dv">5</span>, <span class="dt">p1=</span><span class="fl">0.8</span>, <span class="dt">p2=</span><span class="fl">0.2</span>)</code></pre></div>
<p>How many samples would we need for 90% power?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">power.prop.test</span>(<span class="dt">power=</span><span class="fl">0.9</span>, <span class="dt">p1=</span><span class="fl">0.8</span>, <span class="dt">p2=</span><span class="fl">0.2</span>)</code></pre></div>
<p>Also check out the <a href="https://cran.r-project.org/web/packages/pwr/"><strong>pwr</strong> package</a> which has power calculation functions for other statistical tests.</p>
<table>
<thead>
<tr class="header">
<th>Function</th>
<th>Power calculations for</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>pwr.2p.test()</code></td>
<td>Two proportions (equal n)</td>
</tr>
<tr class="even">
<td><code>pwr.2p2n.test()</code></td>
<td>Two proportions (unequal n)</td>
</tr>
<tr class="odd">
<td><code>pwr.anova.test()</code></td>
<td>Balanced one way ANOVA</td>
</tr>
<tr class="even">
<td><code>pwr.chisq.test()</code></td>
<td>Chi-square test</td>
</tr>
<tr class="odd">
<td><code>pwr.f2.test()</code></td>
<td>General linear model</td>
</tr>
<tr class="even">
<td><code>pwr.p.test()</code></td>
<td>Proportion (one sample)</td>
</tr>
<tr class="odd">
<td><code>pwr.r.test()</code></td>
<td>Correlation</td>
</tr>
<tr class="even">
<td><code>pwr.t.test()</code></td>
<td>T-tests (one sample, 2 sample, paired)</td>
</tr>
<tr class="odd">
<td><code>pwr.t2n.test()</code></td>
<td>T-test (two samples with unequal n)</td>
</tr>
</tbody>
</table>
</div>
<div id="exercise-set-4" class="section level3">
<h3><span class="header-section-number">8.5.3</span> Exercise set 4</h3>
<ol style="list-style-type: decimal">
<li><p>You’re doing a gene expression experiment. What’s your power to detect a 2-fold change in a gene with a standard deviation of 0.7, given 3 samples? (Note - fold change is usually given on the <span class="math inline">\(log_2\)</span> scale, so a 2-fold change would be a <code>delta</code> of 1. That is, if the fold change is 2x, then <span class="math inline">\(log_2(2)=1\)</span>, and you should use 1 in the calculation, not 2).</p></li>
<li><p>How many samples would you need to have 80% power to detect this effect?</p></li>
<li><p>You’re doing a population study looking at the effect of a SNP on disease X. Disease X has a baseline prevalence of 5% in the population, but you suspect the SNP might increase the risk of disease X by 10% (this is typical for SNP effects on common, complex diseases). How many samples do you need to have 80% power to detect this effect, given that you want a statistical significance of <span class="math inline">\(p&lt;0.001\)</span>?</p></li>
</ol>
</div>
</div>
<div id="tidying-models" class="section level2">
<h2><span class="header-section-number">8.6</span> Tidying models</h2>
<p>We spent a lot of time in <a href="r-tidy.html">other lessons</a> on <em>tidy <strong>data</strong></em>, where each column is a variable and each row is an observation. Tidy data is easy to filter observations based on values in a column (e.g., we could get just adult males with <code>filter(nha, Gender==&quot;male&quot; &amp; Age&gt;=18)</code>, and easy to select particular variables/features of interest by their column name.</p>
<p>Even when we start with tidy <em>data</em>, we don’t end up with tidy <em>models</em>. The output from tests like <code>t.test</code> or <code>lm</code> are not data.frames, and it’s difficult to get the information out of the model object that we want. The <a href="https://github.com/dgrtwo/broom"><strong>broom</strong> package</a> bridges this gap.</p>
<div class="figure">
<img src="img/broom.jpg" />

</div>
<p>Depending on the type of model object you’re using, broom provides three methods that do different kinds of tidying:</p>
<ol style="list-style-type: decimal">
<li><code>tidy</code>: constructs a data frame that summarizes the model’s statistical findings like coefficients and p-values.</li>
<li><code>augment</code>: add columns to the original data that was modeled, like predictions and residuals.</li>
<li><code>glance</code>: construct a concise <em>one-row</em> summary of the model with information like <span class="math inline">\(R^2\)</span> that are computed once for the entire model.</li>
</ol>
<p>Let’s go back to our linear model example.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Try modeling Testosterone against Physical Activity, Age, and Gender.</span>
fit &lt;-<span class="st"> </span><span class="kw">lm</span>(Testosterone~PhysActive+Age+Gender, <span class="dt">data=</span>nha)

<span class="co"># See what that model looks like:</span>
<span class="kw">summary</span>(fit)</code></pre></div>
<p>What if we wanted to pull out the coefficient for Age, or the P-value for PhysActive? It gets pretty gross. We first have to <code>coef(summary(lmfit))</code> to get a matrix of coefficients, the terms are still stored in row names, and the column names are inconsistent with other packages (e.g. <code>Pr(&gt;|t|)</code> compared to <code>p.value</code>). Yuck!</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">coef</span>(<span class="kw">summary</span>(fit))[<span class="st">&quot;Age&quot;</span>, <span class="st">&quot;Estimate&quot;</span>]
<span class="kw">coef</span>(<span class="kw">summary</span>(fit))[<span class="st">&quot;PhysActiveYes&quot;</span>, <span class="st">&quot;Pr(&gt;|t|)&quot;</span>]</code></pre></div>
<p>Instead, you can use the <code>tidy</code> function, from the broom package, on the fit:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(broom)
<span class="kw">tidy</span>(fit)</code></pre></div>
<p>This gives you a data.frame with all your model results. The row names have been moved into a column called <code>term</code>, and the column names are simple and consistent (and can be accessed using <code>$</code>). These can be manipulated with dplyr just like any other data frame.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tidy</span>(fit) %&gt;%<span class="st"> </span><span class="kw">filter</span>(term!=<span class="st">&quot;(Intercept)&quot;</span>) %&gt;%<span class="st"> </span><span class="kw">select</span>(term, p.value)</code></pre></div>
<p>Instead of viewing the coefficients, you might be interested in the fitted values and residuals for each of the original points in the regression. For this, use <code>augment</code>, which augments the original data with information from the model. New columns begins with a <code>.</code> (to avoid overwriting any of the original columns).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Augment the original data</span>
<span class="kw">augment</span>(fit) %&gt;%<span class="st"> </span>head

<span class="co"># Plot residuals vs fitted values for males, </span>
<span class="co"># colored by Physical Activity, size scaled by age</span>
<span class="kw">augment</span>(fit) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(Gender==<span class="st">&quot;male&quot;</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(.fitted, .resid, <span class="dt">col=</span>PhysActive, <span class="dt">size=</span>Age)) +<span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p>Finally, several summary statistics are computed for the entire regression, such as <span class="math inline">\(R^2\)</span> and the F-statistic. These can be accessed with <code>glance</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">glance</span>(fit)</code></pre></div>
<p>The <strong>broom</strong> functions work on a pipe, so you can <code>%&gt;%</code> your model directly to any of the functions like <code>tidy()</code>. Let’s tidy up our t-test:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">t.test</span>(AlcoholYear~RelationshipStatus, <span class="dt">data=</span>nha)
<span class="kw">t.test</span>(AlcoholYear~RelationshipStatus, <span class="dt">data=</span>nha) %&gt;%<span class="st"> </span><span class="kw">tidy</span>()</code></pre></div>
<p>…and our Mann-Whitney <em>U</em> test / Wilcoxon rank-sum test:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">wilcox.test</span>(AlcoholYear~RelationshipStatus, <span class="dt">data=</span>nha)
<span class="kw">wilcox.test</span>(AlcoholYear~RelationshipStatus, <span class="dt">data=</span>nha) %&gt;%<span class="st"> </span><span class="kw">tidy</span>()</code></pre></div>
<p>…and our Fisher’s exact test on the cross-tabulated data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">xtabs</span>(~Gender+Diabetes, <span class="dt">data=</span>nha) %&gt;%<span class="st"> </span><span class="kw">fisher.test</span>()
<span class="kw">xtabs</span>(~Gender+Diabetes, <span class="dt">data=</span>nha) %&gt;%<span class="st"> </span><span class="kw">fisher.test</span>() %&gt;%<span class="st"> </span><span class="kw">tidy</span>()</code></pre></div>
<p>…and finally, a logistic regression model:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># fit the model and summarize it the usual way</span>
glmfit &lt;-<span class="st"> </span><span class="kw">glm</span>(Insured~Race, <span class="dt">data=</span>nha, <span class="dt">family=</span>binomial)
<span class="kw">summary</span>(glmfit)

<span class="co"># tidy it up!</span>
<span class="kw">tidy</span>(glmfit)

<span class="co"># do whatever you want now</span>
<span class="kw">tidy</span>(glmfit) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(term!=<span class="st">&quot;(Intercept)&quot;</span>) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">logp=</span>-<span class="dv">1</span>*<span class="kw">log10</span>(p.value)) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(term, logp)) +<span class="st"> </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>) +<span class="st"> </span><span class="kw">coord_flip</span>()</code></pre></div>
<p>Check out some of the other <a href="https://cran.r-project.org/web/packages/broom/index.html"><strong>broom</strong> vignettes on CRAN</a>, and also check out the <a href="http://bioconductor.org/packages/release/bioc/html/biobroom.html"><strong>biobroom</strong> package</a> on bioconductor for turning bioconductor objects and analytical results into tidy data frames.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="reproducible-research.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="surival-analysis.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/07-stats.Rmd",
"text": "Edit"
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
